.github/workflows/release.yml

```diff
@@ -0,0 +1,70 @@
+# Publishes official python package at https://pypi.org/project/ceph-nvmeof/
+# This is official release, and will be pushed to pypi when a new tag is pushed in repo.
+# Usage: pip install ceph-nvmeof==1.2.15
+
+# Also publishes dev python package at https://test.pypi.org/p/ceph-nvmeof
+# This is for dev releases, and will be pushed to test pypi on all merges to "devel" branch.
+# Usage: `pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple ceph-nvmeof==1.2.15`
+
+name: Publish release to PyPI
+on:
+  push:
+    tags:
+      - '*'
+    branches:
+      - 'devel'
+  release:
+    types:
+      - created
+  workflow_dispatch:
+    inputs:
+      python_version:
+        description: 'Python Version'
+        required: false
+        type: string
+      pdm_version:
+        description: 'PDM Version'
+        required: false
+        type: string
+      pypi_repo:
+        description: 'Push to test.pypi (dev) or pypi (prod)'
+        required: true
+        default: 'dev'
+        type: choice
+        options:
+          - dev 
+          - prod-and-dev
+
+env:
+  DEFAULT_PYTHON: ${{ inputs.python_version || '3.9' }}
+  DEFAULT_PDM: ${{ inputs.pdm_version || '2.7.4' }}
+
+jobs:
+  pypi-publish:
+    name: Upload release to PyPI
+    runs-on: ubuntu-latest
+    permissions:
+      id-token: write
+    steps:
+      - uses: actions/checkout@v3
+      - uses: pdm-project/setup-pdm@v3
+        with:
+          python-version: ${{env.DEFAULT_PYTHON}}
+          version: ${{env.DEFAULT_PDM}}
+      - name: Sync Dependencies 
+        run: pdm sync -v --no-isolation --no-self --no-editable
+      - name: Compile Protocol Buffers
+        run: pdm run protoc
+      - name: Set current date as env variable
+        run: echo "NOW=$(date +'%Y%m%d%H%M%S')" >> $GITHUB_ENV
+      # - name: [debug] Set dev version number    # For debugging CI failures (to avoid messing dev releases)
+      #   run: |
+      #     NEW_VERSION="$(pdm show --version).dev$NOW"
+      #     sed -i "s/^version = \".*\"/version = \"${NEW_VERSION}\"/" pyproject.toml
+      - name: Build package
+        run: pdm build --config-setting="--build-number=$NOW" --no-sdist
+      - name: Publish package distributions to PyPI
+        run: pdm publish --no-build
+        if: startsWith(github.ref, 'refs/tags/') || (inputs.pypi_repo && inputs.pypi_repo == 'prod-and-dev')
```

Comment:
I'd simplify this. When do you want each build type to happen?

* the prod package publishing should ONLY happen as part of the Github "Create release" process, which internally may involve creating a tag (but not necessarily). Hence, the best option would be to trigger [the job on `release: published` events](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release).
  * If a manual prod amendment is required, the `workflow_dispatch` would do the trick.
* Test packages may happen (depending on whether PyPI has some storage quota or not):
  * On-demand (`worklfow_dispath`),
  * Whenever a PR is opened (`pull_request_target` against `devel`),
  * Whenever a PR is merged (`push` against `devel`).

Reply:
@epuertat, thank you so much for the reviews! Sorry for the delay in updating the PR.
I looked into pypi storage limits and there is indeed a storage limit for a project of 10GB/project: https://pypi.org/help/#:~:text=project%20size%20limit%20(10.0%20GiB%20by%20default

So I have added an input in `workflow_dispatch` called "ref" which takes in which branch/commit to checkout before building. This doesn't cover pull requests made from forks, let me know if I should add an input to accept repo.